File: QtTestBrowser/xmlccnxrequest.cpp
#include "xmlccnxrequest.h"

#include <QUrl> 
#include <QDebug>
#include "ccnximpl.h"

XMLCCNxRequest :: XMLCCNxRequest(QObject *parent): QObject(parent){
	
}

XMLCCNxRequest :: ~XMLCCNxRequest(){

	if(m_ccnxImpl != NULL){
		m_ccnxImpl->stop();
		m_ccnxImpl->wait();
	}	
}

void XMLCCNxRequest :: send(QString type, QString url, QString params, bool async, bool responseFromServer){

	// validate type, url and params
	if(url.startsWith("/"))
		url += "ccnx:/";

	if(paramsIsValid(params)){

		QUrl request_url(url);
		outgoingData = params.toUtf8();
		qDebug() << "Outgoing Data:" << outgoingData;
		//qDebug() << "XMLCCNxRequest --> responseFromServer" << responseFromServer;
		if(async){
			
			// try to send interest async
			m_ccnxImpl = new CCNxImpl(request_url, outgoingData, responseFromServer);

			connect(m_ccnxImpl, SIGNAL(contentReadyRead(QByteArray, QString)), this,
				SLOT(receive(QByteArray, QString)));
			connect(m_ccnxImpl, SIGNAL(interestExpressed(QUrl)), this,
				SLOT(sent_slot(QUrl)));
			connect(m_ccnxImpl, SIGNAL(error(QString)), this,
				SLOT(error_slot(QString)));
			connect(m_ccnxImpl, SIGNAL(finished()), this,
				SLOT(finished()));

			m_ccnxImpl->start();

			/* async operation */
			
		}else{

			// try to send interest sync
			m_ccnxImpl = new CCNxImpl(request_url, outgoingData, responseFromServer);

			connect(m_ccnxImpl, SIGNAL(contentReadyRead(QByteArray, QString)), this,
				SLOT(receive(QByteArray, QString)));
			connect(m_ccnxImpl, SIGNAL(interestExpressed(QUrl)), this,
				SLOT(sent_slot(QUrl)));
			connect(m_ccnxImpl, SIGNAL(error(QString)), this,
				SLOT(error_slot(QString)));
			connect(m_ccnxImpl, SIGNAL(finished()), this,
				SLOT(finished()));

			m_ccnxImpl->start();	

			/* sync operation */							
		}			
	}
}


void XMLCCNxRequest :: receive(QByteArray content, QString received_type){
	qDebug() << "AJAX request receive response";

	QString data(content);
    this->data += data;
}


void XMLCCNxRequest :: sent_slot(QUrl url){
	qDebug() << "AJAX request sent:" << url.toString();

	QMap<QString, QVariant> response;
	response["statusCode"] = QVariant(Sent);
	response["data"] = QVariant("Interest Has Sent");
 
	emit sent(response);
}


void XMLCCNxRequest :: error_slot(QString errorInfo){

	qDebug() << "AJAX request error:" << errorInfo;
	QMap<QString, QVariant> response;
	response["statusCode"] = QVariant(Error);
	response["data"] = QVariant(errorInfo);
 
	emit error(response);
}


void XMLCCNxRequest :: finished(){

	QMap<QString, QVariant> response;
    //QString datanew;
	response["statusCode"] = QVariant(Success);
    /*for(int i = 0; i < data.length()/2;i++){
       datanew[i] = data[i];
    }*/
    response["data"] = QVariant(data);

	emit finish(response);
}

void XMLCCNxRequest :: stopRequest(){
	m_ccnxImpl->stop();
}


bool XMLCCNxRequest :: paramsIsValid(QString params){

	qDebug() << "Params:" << params << " " << QString::compare(params.trimmed(), "");
	if(QString::compare(params.trimmed(), "") == 0)
		return true;

	QRegExp equation("[^=]*=[^=]*");
 	QStringList paramsList = params.split("&");
	QList<QString>::iterator it = paramsList.begin(), itEnd = paramsList.end();	
	for(; it != itEnd; it++){
		if(!equation.exactMatch(*it)){	
			QMap<QString, QVariant> response;
			response["statusCode"] = QVariant(Error);
			response["data"] = QVariant("Please check your params.");	
			emit error(response);
			qDebug() << "AJAX Params Error.";
			return false;
		}
	}

	return true;
}

File: QtTestBrowser/xmlccnxrequest.js
/*
 * Please include the javascript library <xmlccnxrequest.js> when using the following methods.
 *
 * USAGE Normal:
 * 	var xmlCCnxrequest = new XMLCCNxRequest();
 * 
 * 	xmlCCNxRequest.success = function(data){
 * 		//processure
 * 	};
 * 	xmlCCNxRequest.error = function(data){
 * 		//processure
 * 	};	
 * 
 * 	xmlCCNxRequest.send("POST", "ccnx://bupt/get.jsp", "t="+Math.random(), true, true);
 * 	xmlCCNxRequest.post("ccnx://bupt/get.jsp", "t="+Math.random(), true, true);
 * 	xmlCCNxRequest.get("ccnx://bupt/get.jsp?t="+Math.random(), true, true);
 * 
 * USAGE AJAX:
 * 
 * 	AJAX_CCNx({		
 * 		url: "ccnx://get.jsp",
 * 		type: "GET",
 * 		params: "t="+Math.random(),
 * 		sent: function(data){
 * 			//processure				
 * 		},
 * 		success: function(data){
 * 			//processure
 * 		},
 * 		error: function(data){
 * 			//processure
 * 		}				
 * 	});
 *
 * Tips: The parameter <params> will be deglected when using GET method <send>.
 */


function XMLCCNxRequest(){
	
	this.xmlCCNxRequest = ccnxRequest.createXMLCCNxRequest();

	this.beforeSend = function(){};
	this.success = function(data){};
	this.error = function(data){};
	this.sent = function(data){};
	this.async = true;
	this.responseFromServer = true;
	this.url = "";
	this.type = "GET";
	this.dataType = "text";
	this.params = "";

	this.send = function(type, url, params, async, responseFromServer){

		this.xmlCCNxRequest.sent.connect(this.sent);
		this.xmlCCNxRequest.error.connect(this.error);
		this.xmlCCNxRequest.success.connect(this.success);
		this.type = type;
		this.url = url;
		this.params = params;
		this.async = async;	
		this.responseFromServer = responseFromServer;
		this.beforeSend();	
		this.xmlCCNxRequest.send(type, url, params, async, this.responseFromServer);

	};

	this.send_inner = function(){

		this.xmlCCNxRequest.sent.connect(this.sent);
		this.xmlCCNxRequest.error.connect(this.error);
		this.xmlCCNxRequest.finish.connect(this.success);	
		this.beforeSend();
		this.xmlCCNxRequest.send(this.type, this.url, this.params, this.async, this.responseFromServer);
	};

	this.get = function(url, async, responseFromServer){

		this.xmlCCNxRequest.sent.connect(this.sent);
		this.xmlCCNxRequest.error.connect(this.error);
		this.xmlCCNxRequest.success.connect(this.success);
		this.type = "GET";
		this.url = url;
		this.async = async;	
		this.responseFromServer = responseFromServer;
		this.beforeSend();	
		this.xmlCCNxRequest.send(type, url, params, async, this.responseFromServer);
	};

	this.post = function(url, params, async, responseFromServer){

		this.xmlCCNxRequest.sent.connect(this.sent);
		this.xmlCCNxRequest.error.connect(this.error);
		this.xmlCCNxRequest.success.connect(this.success);
		this.type = "POST";
		this.url = url;
		this.params = params;
		this.async = async;	
		this.responseFromServer = responseFromServer;
		this.beforeSend();	
		this.xmlCCNxRequest.send(type, url, params, async, this.responseFromServer);
	}
}


function AJAX_CCNx(request){

	var ccnx_ajax = new XMLCCNxRequest();

	for (e in request){
		switch(e){
			case "beforeSend":
				ccnx_ajax.beforeSend = request.beforeSend
				break

			case "success":
				ccnx_ajax.success = request.success
				break

			case "error":
				ccnx_ajax.error = request.error
				break

			case "sent":
				ccnx_ajax.sent = request.sent
				break

			case "async":
				ccnx_ajax.async = request.async
				break

			case "url":
				ccnx_ajax.url = request.url
				break

			case "type":
				ccnx_ajax.type = request.type
				break

			case "dataType":
				ccnx_ajax.dataType = request.dataType
				break

			case "params":
				ccnx_ajax.params = request.params
				break

			case "responseFromServer":
				ccnx_ajax.responseFromServer = request.responseFromServer
				break

			default:
				break

		}
	}

	ccnx_ajax.send_inner();	
} 


